private_sharing/api_views.py

```diff
@@ -226,7 +227,31 @@ def post(self, request, format=None):
         return Response("success")
 
 
-class ProjectFileDirectUploadView(ProjectFormBaseView):
+class SaveDataTypesMixin(object):
+    """
+    Mixin to do some final verification of datatypes to be associated with the datafile
+    and to do the actual association.
+    """
+
+    @property
+    def good_datatypes(self):
+        """
+        Verifies that the supplied datatypes are valid.  Returns boolean.
+        """
+        datatypes_ids = set(self.project.datatypes.all().values_list('id', flat=True))
+        return self.form.cleaned_data["datatypes"].issubset(datatypes_ids)
+
+    def save_datatypes(self, data_file):
+        """
+        Saves the provided datatypes in the ProjectDataFile instance.
+        """
+        for datatype_id in self.form.cleaned_data["datatypes"]:
+            datatype = self.project.datatypes.get(id=datatype_id)
+            data_file.datatypes.add(datatype)
+        data_file.save()
```

Comment:
Can you just add the datatypes, then make sure to save the model after calling this function (not before)?

Reply:
The documentation explicitly states that you cannot associate objects with a manytomany field until the object has been saved; it must be done separately.